### DISCUSSION TITLE
[ARCHIVED] Низкоуровневое именование методов/классов/констант

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#12] [Archive]
> handleClick vs clickHandle?
> createUser vs user.create vs userCreate?

---

Методология скорее всего будет рекомендовать нейминг `{entity}-{action}` (н-р: `user-create`, `post-view`, ...) в большинстве случаев

*Но @sergeysova выступает с инициативой, чтобы дополнительно рекомендовать такой подход наименования и не только при структуре*
> Т.е. и при наименовании методов / обработчиков

**Comment by sergeysova** [comment]
There is a useful pattern to follow when naming functions:

```
namespace? + high context (HC) + low context? (LC) + action (A)
```

Take a look at how this pattern may be applied in the table below.

| Name                 | Namespace | High context (HC) | Low context (LC) | Action (A) |
| -------------------- | --------- | ----------------- | ---------------- | ---------- |
| `userGet`            |           | `User`            |                  | `Get`      |
| `userMessagesGet`    |           | `User`            | `Messages`       | `Get`      |
| `clickOutsideHandle` |           | `Click`           | `Outside`        | `Handle`   |
| `shouldMessageShown` | `should`  | `Message`         |                  | `Shown`    |

> **Note:** The order of context affects the meaning of a variable. For example, `shouldUpdateComponent` means _you_ are about to update a component, while `shouldComponentUpdate` tells you that _component_ will update on itself, and you are but controlling when it should be updated.
> In other words, **high context emphasizes the meaning of a variable**.

#### Why action goes the latest?

When you write code in IDE, you have an "code intelligency" or autosuggestion. If you type action first "get<tab>", you'll see all function in project that has "get" action, and you need to type High context and Low context to reduce amount of methods.
But if you first type High context you can see all functions that can operate with that context and with lowest contexts, then you can narrow the search typing the Low context name, and now you see all functions you can do with that context and nothing more.


https://gist.github.com/sergeysova/9f6dc24e21efe5f5bb4f9878216c2ce9

**Reply by ilyaagarkov** [reply]
Я бы подобное обсудил с носителем языка.

Так как мои  знания языка говорят что userGet делает сам код менее читаемым, так как мы нарушаем грамматику английского языка.  UserGet говорит о том что юзер что-то получает, а не то что мы получаем юзера.
Название функции должно говорить, что она делает. Чем ближе код к естественному языку тем лучше.

Аргумент про IDE имеет место быть, но все таки код мы пишем в первую очередь для людей, а не для IDE.
Да это улучшит "code intelligency", но написав `get` разработчик и так уже знает о каком контексте(HC) идет речь, и в состоянии сам набрать его.

**Reply by sergeysova** [reply]
> userGet делает сам код менее читаемым, так как мы нарушаем грамматику английского языка. UserGet говорит о том что юзер что-то получает, а не то что мы получаем юзера.
Название функции должно говорить, что она делает. Чем ближе код к естественному языку тем лучше.

Проблема в том, что невозможно писать код на английском. И в какой-то момент тебе придется выбрать между конвенцией и английским языком.

> все таки код мы пишем в первую очередь для людей, а не для IDE
Да, как и конвенции нужны людям.

Если в проекте нет конвенции именования, то мы получаем ситуацию, когда два разных разработчика назовут функции катастрофически по разному.

Ещё пример: `shouldComponentUpdate` и `shouldUpdateComponent`, без конвенции непонятна разница и придется добавлять кучу "мусорных" слов и артиклей для указания принадлежности, как того требует английский(чего мы и так не делаем в большинстве случаев).
С конвенцией всегда понятно, что важнее и как читать такие имена.

Я имею ввиду, что в коде мы НЕ ПИШЕМ на английском, хоть и используем английские слова. Фактически, конвенция выступает своего рода грамматикой для разработчиков, как понимать друг друга в рамках ограничений языка программирования, отсутствия знаков препинания и смысловых акцентов.

**Comment by ilyaagarkov** [comment]
Почему данный вопрос вообще рассматривается в рамках feature-sliced? 
По-моему тут нарушен принцип единой отвественности.

**Reply by sergeysova** [reply]
С одной стороны кажется, что это out of scope.
Но с другой стороны, у нас всё равно будут рекомендации по именованию фич и entities.

Если получится прийти к общему пониманию именования, то можно добавить **рекомендацию** по именованию сущностей в проекте, чтобы разработчикам использующим методологию не нужно было об этом думать и они сразу могли взять готовый подход

**Reply by azinit** [reply]
@ilyaagarkov, https://github.com/feature-sliced/wiki/discussions/12#discussioncomment-455272

**Comment by azinit** [comment]
`WARNING:` С этого сообщения и родилась эта ~~холиварня~~ тема
---

> @karina-drummer 
> Попробую сорировать методы API по типу запроса, но именовать типы согласно CRUD

**Reply by azinit** [reply]
<details>
<summary>
Оччч большая история сообщений)
</summary>

> @martis-git 
> Ну это то да, мы тож всегда стараемся требовать ,чтоб на бэке проставляли operationId согласно CRUD
> 
> И тогда получаются операции аля:
> - getTasks
> - createTask
> - getTask
> - updateTask
> - deleteTask
>
> И с таким уже проще гораздо работать (когда сгенерируется через openapi-generator какой-нить)
---
> @sergeysova 
> Лучше обратно именовать
> tasksList
> taskCreate
> taskGet
> ...
---
> @karina-drummer 
> А если полностью по CRUD?
>
> tasksRead
> taskCreate
> taskRead
---
> @karina-drummer 
> Я бы правда в таком случае дописывала в конце "Request" для readability, хоть это и муторно
---
> @sergeysova 
> Круд бесполезен на чем нибудь более сложном
---
> @martis-git 
> Чую еще придется обсудить этот вопрос интенсивно)
>
> Я наоборот старался именовать так, чтобы  было как в CRUD репозиториях (на том же бэке)
>
> в том числе так более ясно как будто (т.е. у тебя конкретный эндпоинт == какое-то CRUD действие )
---
> @martis-git 
> Ну вот да, хотелось бы увидеть примеры проектов, где вообще нет круда)
> Понятно что там и другие методы появляются
>
> Но интересно тогда даже, а зачем само действие/метод сувать в конец
> 
> Т.к. например обращаясь к какому то инстансу АПИшки (TaskApi, BlogPostApi), в том числе хочется чтоб по предлагаемому автокомплиту было понятно что можно от сервера запросить (не заходя в сваггер  и ища похожий урл)
> ![image](https://user-images.githubusercontent.com/42924400/110476927-9dee9900-80f3-11eb-8dc1-a8609f128021.png)
---
> @martis-git 
> Не могу сказать, что у нас идеальная АПИшка генерируется (не забываем про кодогенерацию из сваггер схемы)
> **Но в целом CRUD концепция все равно наслаивается**
>
> И для разраба лучше обычно всегда чтоб метод в начале был
---
> @martis-git 
> Т.е. в том числе вопрос (который ты тоже помню кидал в гисте), как понятнее и лучше разрабу:
> toggleSection
> или
> sectionTogle
---
> @sergeysova 
> Восстановление пароля это какое действие круд?
---
> @martis-git 
> Ну это скорее всего какой-нить PUT/PATCH метод был бы
> 
> Я не хочу прям строгую связку с CRUD делать, само собой)
---
> > @martis-git
> > Т.е. в том числе вопрос (который ты тоже помню кидал в гисте), как понятнее и лучше разрабу:
>
> @sergeysova 
> Понятнее то, в чем он разобрался и принял для себя
---
> @martis-git 
> Но когда название метода (будь то CRUD или другое) начинается не с самого "действия" (get/reset/create/update/...) а с сущности - ну хизе хизе))
---
> @sergeysova 
> Crud и rest в реальных приложениях не работает

> @sergeysova 
> [Проблема автокомплита в IDE, когда ты по простому запросу получаешь простыню лишних методов]
---
> @martis-git 
> А что нам мешает заставить бэк написать нормально сваггер схему, чтобы у нас разделялась апишка для сущностей?)
>
>  *(MaterialsApi, TasksApi, IssueApi, BlogPostApi и т.д.)*
![image](https://user-images.githubusercontent.com/42924400/110477230-f9b92200-80f3-11eb-85d6-ecb7b8662290.png)
---
> @martis-git 
> И тип когда ты обращаешься к BlogPostApi - там даже если будет так много методов, то они так и ли иначе с блог-постами и связаны)
---
> > @martis-git
> > И тип когда ты обращаешься к BlogPostApi - там даже если будет так много методов, то они так и ли иначе с блог-постами и связаны)
>
> @sergeysova 
> Это не работает
---
> @martis-git 
> Т.е. тут главная проблема возможная то что вроде как разрабы привыкли писать в рамках {action}{Entity}, а тут мы даже на уровне структуры им что то навязываем)
>
> Но в целом да - я это как раз и говорю
> 
> Будет идеально, если это на нашу методологию не повлияет
---
> @sergeysova 
> Не привыкли)
> Напомни им про объект.действие()
> И сразу станет на места
---
> @martis-git 
> "Упростить весь процесс" - именно то чего хочется достичь)
>
> Как можно меньше нюансов чтоб было как раз
> Или чего-то оч навязанного
---
> @martis-git 
> оч редко когда ты из автокомплита можешь выбрать что-то точное
Благо все IDEшки даже если у тебя эндпоинт называется "getTasksLogs", ты можешь ввести "tasklogs" - и ты в том числе и получишь нужное
>
> Т.е. мы как будто не тем путем пытаемся решить проблему
Если изначально боль имеено "большой автокомплит, IDE не может угадать мой поток мыслей - "сначала сущность, потом действие"" - то как будто это проблема IDE
---
> @sergeysova 
> То же самое с сортировкой сущностей по алфавиту
---
> @martis-git 
> Т.е. хочешь сказать что нет тендеции, что большинство разрабов будут писать handleClick вместо clickHandle?
> 
> Тип даже в дискуссиях другого масштаба обычно везде как-то стараются базовую операцию (handle/on в частности) ставить в начале, а потом уже событие/сущность
>
> Возможно все сообщество ошибается и надо делать как ты, но тогда оч все же интересно - какую проблему решает твой подход помимо автокомплита)
>
> Я пытаюсь понять идеологически в том числе... И чет хз пока)
---
> @sergeysova 
> Предлагаю обратить внимание на причины, почему ты споришь)
> Не связано ли это ТОЛЬКО с привычками?
---
> @martis-git 
> Ну вот да, я как раз таки не то что хочу "лишь бы поспорить, и  свою парадигму доказать всем вам"
>
> А именно твой вижн проблемы пытаюсь до конца понять
>
> Тип почему именно так
---
> @sergeysova 
> группировка по сущностям
> очевидно, что так проще посмотреть все методы?
> ты ищешь метод который делает что-то с сущностью
> а не ЧТО он делает
---
> @martis-git 
> "Хотим ли мы легкого порога входа, чтоб каждый примерно так уже писал раньше, но не знал как правильно"
> Или хотим "Навязать и убедить всех как нужно правильно, забивая на текущие практики и привычки разрабов"
>
>
> Мб немного с эмоциональной окраской получилось, но я по крайней мере пока так проблему вижу)
---
> @sergeysova 
> это не взаимоисключающие подходы
---
> @sergeysova 
> чтобы “реально пользовались” давай будем писать как в го
> сколько миллионов пользователей у го?
> как тебе го как язык с идеологией?
>
> это вопрос целевой аудитории
> ты хочешь взять всех
> или только тех, кому важна структура и системность?
---
> @sergeysova 
>  Представь: ты пишешь код в рамках методологии и у тебя проблемы возникают раз в 10 реже, просто потому, что другие люди продумали решение многих проблем.
>
>
> Нейминг от “общего к частному” это такой же способ решить проблемы, но в этот раз проблемы восприятия.
>
>
> Если же у тебя нет паттерна нейминга, как тогда именовать? Как придется?
> 
> https://github.com/MostlyAdequate/mostly-adequate-guide-ru/blob/master/SUMMARY-ru.md
> Для кого-то фп это максимально понятно
>
---
> @sergeysova 
> Ещё довод:
> Если именовать фичи по моему подходу, то сгруппировать их в кластер можно без поломки именования
![image](https://user-images.githubusercontent.com/42924400/110478268-3b969800-80f5-11eb-8df1-ee3bae3a9cbf.png)
![image](https://user-images.githubusercontent.com/42924400/110478284-3e918880-80f5-11eb-9eb1-4ab480e31d40.png)
>
> Как видишь, нейминг опять возвращается к «от общего к частному». И в итоге получается сломанный подход, где-то в проекте пишут так, а где-то так. Это не система, это не консистентно
>
![image](https://user-images.githubusercontent.com/42924400/110478318-48b38700-80f5-11eb-825d-9f2929bf158c.png)
> Даже с апи. Ты сам приводил такой же пример. И в нем же тоже реверсируешь свой же подход при обобщении
>
> 
![image](https://user-images.githubusercontent.com/42924400/110478367-5537df80-80f5-11eb-99e3-dfff35cf3bdd.png)
> Вот взгляни на всё это и скажи мне, что я предлагаю не системный не консистентный подход
>
>
---
> @martis-git 
> Так забайтил конечно)))

---
> @martis-git 
> Не, я со всем что ты из последнего написал - согласен
> И наверное некоторые частные моменты не стоит разруливать, и в том числе соглашусь, что:
> `features/user/create`
> `features/user/delete`
> и
> `api.users.get`
>
> Выглядят получше аналогов (и этого наверное нам пока хватит)
---
> @sergeysova 
> Так получается что кусок так, а другой кусок так
> С методами и функциями все то же самое
---
> > @sergeysova
> > Так получается что кусок так, а другой кусок так
> > С методами и функциями все то же самое
>
> @martis-git 
> Ну смотри
> Основные моменты вроде как у нас совпадают (что "возможно", что "цель")
> 
> А такие частности - как именовать хендлер - handleClick или clickHandle - думаю пока рил больше в реализацию скатывается)
---
> @martis-git 
> Тип мб и я пойму со временем (в том числе и другие разрабы), что* методы и обработчики надо называть начиная с сущности
> 
> Но нам повезло, что такие моменты методология наша пока не должна разрешать)
> // Возможно потом* что-то появится такое, но это уже будет след. итерация
> 
> И уже хорошо - что мы понимаем целевую аудиторию и ключевые моменты, и думаю пока исходя из этого - можно дальше продолжать работу)
---
> @martis-git 
> А то сейчас к долгому и глубокому срачу может привести, который не факт что нам профит принесет на данном этапе)
> 
> // Если будем обсуждать как называть обработчики
---
> > @martis-git
> > Ну смотри
> > Основные моменты вроде как у нас совпадают (что "возможно", что "цель")
>
> @sergeysova 
> Тут я тоже согласен
>  Поэтому такое лучше не совать в спеку
> Но я бы добавил рекомендацию по именованию
> С пометкой, что так можно именовать все виды сущностей в проекте, если вам подходит
> Но фичи, процессы и другие директории строго рекомендуется именовать по этому подходу
---

</details>

- ограничения CRUD
- проблемы нейминга
- проблемы большого автокомплита и мышления разработчика
- `actionSubject()` vs `subject.action()` vs `subjectAction()`

**Reply by azinit** [reply]
<details>
<summary>
Two hours later...
</summary>


> @karina-drummer 
> ![image](https://user-images.githubusercontent.com/42924400/110479146-279f6600-80f6-11eb-94f9-45385a2c7f54.png)

---
> @karina-drummer 
> Я просто никак не могу понять, зачем в имени каждого метода повторять имя сущности, если можно
> 
> api.skins.read.byId()
> api.skins.read.list()
> api.skins.create.newSkin(
---
> @sergeysova 
> Это в идеале.
> Но такую структуру сделать крайне сложно
---
> @karina-drummer 
> А вот напиши, мне нужен конструктив
---
> @sergeysova 
> Гораздо проще сгенерить из сваггера методы, чем глубоко вложенную структуру
> 
> Сваггер не позволяет вложенность описать
---
> @sergeysova 
> В сваггере есть только operationId и tags
> 
> В итоге имеем
> Tag.operationId
> 
> Тег нужен для группировки и мы его не юзаем в генераторе. А вот в operationId зашиваем всю нужную инфу
---
> @martis-git 
> Угу, и проблемы в том, чтоб operationId был уникальным для всей схемы
---
> > @karina-drummer 
> > Я бы правда в таком случае дописывала в конце "Request" для readability, хоть это и муторно
> 
> @karina-drummer 
> Оукей. Если уж так, то может в этом есть смысл
> В случае с теми же сагами читается как предложение
---
> @sergeysova 
> Fluent очень тяжело сделать
> Особенно с типизацией
---
> > @sergeysova 
> > В итоге имеем
> > Tag.operationId
>
> @karina-drummer 
> А подкиньте хороший пример плиз

---
> @karina-drummer 
> Мы же просто добавляем "Request" в конце имени вообще любого метода
---
> @sergeysova 
> Когда у тебя в начале есть api это смысла не имеет
> Это «мусор» как по мне
---
> @sergeysova 
> [скинул api], @martis-git , сможешь переложить это апи на rest crud
---
> @martis-git 
> Ну моей целью было другое донести / узнать)
>
> Не буду расписывать детали, но самое главное, что для нас обоих session.create > [sessionCreate, createSession]
>
> А как там уже между собой будут sessionCreate и createSession воевать - вопрос на будущее)
---
> @sergeysova 
> Там так и показано
---
> @karina-drummer 
> Я всегда стараюсь в декларативный нейминг, то есть не использовать глаголы насколько это возможно, как минимум. Поэтому у меня встречается какой-нибудь api.candidatesListRequest
---
> @sergeysova 
> Только глаголы нужны
> А мусор не оч
---
> @sergeysova 
> gist.github.com/sergeysova/9f6dc24e21efe5f5bb4f9878216c2ce9
---
---
> Не ребят, я правда задолбался копировать сообщения :DDDD

![image](https://user-images.githubusercontent.com/42924400/110480602-b95ba300-80f7-11eb-8b03-7f2af31df975.png)
![image](https://user-images.githubusercontent.com/42924400/110480642-c4aece80-80f7-11eb-8618-49e7f510849a.png)
![image](https://user-images.githubusercontent.com/42924400/110480661-cb3d4600-80f7-11eb-8297-26d879d391bd.png)
![image](https://user-images.githubusercontent.com/42924400/110480697-d2fcea80-80f7-11eb-9592-6de1b6b5d79b.png)

</details>

- про swagger, автогенерацию и tags + operationIds
- про английский язык, читабельность и "неправильные привычки разработчиков"

**Reply by sergeysova** [reply]
Думаю в следующий раз не надо копировать всё, а только основные доводы)

**Reply by azinit** [reply]
> Думаю в следующий раз не надо копировать всё, а только основные доводы)

Пожалуй, но чет лень было)

> Там вся история сообщений настолько красноречива, что решил оставить как есть)

**Reply by mindyourlifeguide** [reply]
У себя использую `event => entities => effect` :
`getUsersList`
`updateProfileStatus`
 
вполне удобно и читабельно.

**Reply by azinit** [reply]
> Думаю в следующий раз не надо копировать всё, а только основные доводы)

@sergeysova , схлопнул, темы базово расписал)

**Comment by azinit** [comment]
> @mindyourlifeguide 
> У себя использую `event => entities => effect` :
> `getUsersList`
> `updateProfileStatus`
> 
> вполне удобно и читабельно.

Очень похоже на то, что описано [здесь ](https://github.com/kettanaito/naming-cheatsheet#ahclc-pattern) (да и в большинстве источников)

**Comment by azinit** [comment]
[Ref](https://t.me/c/1463227827/164630)

> @sergeysova 
> Как я именую всё в приложении.
> Когда читаю код, я хочу понимать намерения автора, что он хотел выразить кодом.
>
>Если я вижу функцию employeeStatusChange, я хочу четко понимать её назначение “сменить статус сотрудника” или же “узнать в каком состоянии находится процесс изменения сотрудника”.
>
> Если коротко, то использую подход:
> Namespace + HighContext + LowContext + Action
>
> Примеры выше можно показать так (namespace отсутствует): 
> employeeStatusChange
> employee — high context, в контексте чего выполняется действие
> status — low context, нечто принадлежащее родительскому контексту
> change — action, что делаем с этим
> То есть изменяем статус сотрудника.
>
> А вот пример с выяснением текущего статуса процесса изменения сотрудника, может выглядеть так employeeChangeStatus:
employee — high context, всё так же работаем с сотрудником
change — low context, но здесь уже указываем что работаем с сущностью change принадлежащей сотруднику
status — action, выступает как глагол
>
> Таким образом, можно четко обозначить назначение функции, не добавляя дополнительных слов. Получается коротко и понятно, при условии, что конвенцию понимают все в команде.
>
> А вот как быть БЕЗ конвенции? Как отличить эти функции не превращая имя в месиво вроде readStatusOfEmployeeChanging, которое не отсортировать по имени, потому что тогда сгруппируются все функции начинающиеся с read*, вместо группирования по сущности employee.
>
>P.S.
> Если переложить этот подход на точечную нотацию, получится вполне себе вменяемая объектная модель:
employeeChangeStatus -> employee.change.status()
>
> А вот readStatusOfEmployeeChanging уже не перекладывается без полной смены порядка слов, что не очень приятно. Ну и важные слова employee и change теряются в середине и конце названия.

**Reply by azinit** [reply]
[Ref](https://t.me/c/1463227827/164673)

## Естественный и Формальный языки

> **Vitaly**
> Есть неплохое правило, чтобы понять что происходит, описать действие, начни с глагола
Мы же не говорим счет банковской карты изменить, обычно изменить счет банковской карты

> @sergeysova 
> Мы не пишем код в соответствии с семантикой естественного языка

> **Vitaly**
> Мы пишем код вплотную взаимодействую со стекхолдерами и стараемся говорить на одном языке, чтобы всем лицам в процессе было понятно и можно было легко вносить изменения и поддерживать систему

**Reply by azinit** [reply]
[Ref](https://t.me/c/1463227827/164709)

## Конвенции

> **Natalia**
>
> А есть ли способ жить без конвенций? Ну точнее без конвенция помогающих читать что-то непонятное, но с конвенцией как создавать автоматически читабельные имена. Для меня employeeChangeStatus: все равно непонятная хрень. У работника может быть статус и он может меняться, но у работника не может быть изменений, у которых еще и статус есть. Если это замена одного работника на другого, то это статус его вакансии/position (рабочее место в смысле) и выходит на уровень выше, если изменение чего-либо у сотрудника, то у этого чего-либо есть имя и employeeSomethingChange. Проблема в том, что я начинаю об этом думать, и если я сомневаюсь, я лезу в доку читать что мы там наконвентили, потом пытаюсь расшифровать что имелось в виду, хотя должна была непосредственно решать задачу.
> 
> А еще вопрос в тему архитектур и дробления (если есть мысли буду рада, если в канале напишешь, в чате вряд ли успею заметить ответ). Как ты проводишь грань между простотой и common solution. Вот например, можно сделать одну карточку, которая рендерит 3 разных вида и такие и сякие, а можно сделать три компонента, для каждой карточки отдельно. И вроде оптимально сделать одну карточку на все случаи жизни, а с другой стороны может лучше (для редакторов контента и вообще) наоборот разделить эти компоненты, даже если половина кода там будет повторяться, чтобы структура каждой карточки была чистой вместо одного монстра-карточки.

> @sergeysova 
> К сожалению естественный язык слишком многозначен, каждое слово имеет несколько значений в разных контекстах, зависит даже от того, кто его произносит и слушает. Код это формальное описание логики, то есть мы намеренно лишаем естественный язык его многозначности, чтобы было проще понимать разным людям.
>
> Когда все договорились, что for это цикл, нет никаких проблем понимания кода. Аналогичная ситуация обстоит и с паттернами программирования.
>
> Такие подходы в именовании помогают ограничить фантазию разработчиков и внести формальные правила, которые все одинаково понимают.
>
> Я согласен, что в случае с employeeChangeStatus ситуация выглядит очень натянуто, но это реальный кейс, стоит наверное имена получше подобрать. А кейс такой: информация о сотруднике после изменения должна распространиться по множеству других сервисов, соответственно у нас есть сам метод изменения и метод проверки статуса этого изменения(завершено/в процессе/отвалилось)

**Reply by azinit** [reply]
[Ref](https://t.me/c/1463227827/164772)

## Восприятие и Привычка

> **Bohdan**
> Тут, как по мне, деление именования на фичи больше вредит восприятию, чем помогает.
>
> *съесть кашу вечером*
> *кашу вечером съесть*

> @sergeysova 
> Формальный язык != естественному
>
> Ты не можешь на естественном языке писать код.
Рубисты пытались уже

> **Aleksandr**
> Если действий над Кашей много то удобно же

> **Bohdan**
> Ну ок)
> Если так удобнее - без вопросов)
> Для меня глагол останется на своём месте)
> ...
> Я уже выше писал - если тебе норм, то ок,  я останусь на старой конвенции.

> @sergeysova 
> Да я ж тебя не заставляю)
> Просто лично мне крайне неприятно, что в списке методов все сгруппировано не по сущностям, а по методам.
То есть куча create, потом куча edit и т.д

> **Bohdan**
> Я понимаю, но это уже переходит в фючеслайс всего
> *Методология должно помогать мыслить, а не перестраивать ход моих мыслей под себя*

> **Arthur**
> когда люди используют методологию, всем пофиг кто и как мыслил... прежде надо понять той код и бизнес логику)
> А вообще лучше хоть такой конвенш чем никакой, а то куда не глянь творческий беспорядок))

> **Bohdan**
> на github мне даже ссылочки скинули)
> https://github.com/kettanaito/naming-cheatsheet#ahclc-pattern

**Reply by azinit** [reply]
Я думаю, это обсуждение лишний раз показало - что подобная конвенция - лишь один из способов разработчиков **договориться внутри проекта**, и подходит он - далеко не всем

А потому наименования `handleOutsideClick` vs `clickOutsideHandle` - выносятся за рамки *core-concepts* методологии
> Максимум - такие подходы могут *"рекомендоваться"*, но выбор конкретной конвенции в данном случае - будет зависеть чисто от архитектора проекта и его команды

**Reply by azinit** [reply]
При этом ясно - что эта конвенция применяется и в подходе **в какой-то мере**

> например в [доке по Public API](https://github.com/feature-sliced/wiki/blob/bdb31ab06494932e6c75c24f14881777532d2d5e/concepts/public-api.md)
![image](https://user-images.githubusercontent.com/42924400/112417070-e3d46f80-8d37-11eb-8d4f-0dc4dea79c5c.png)

*Другой вопрос - насколько готов разработчик ~~упороться~~ доводить этот подход до абсолюта?*

**Reply by akaia-shadowfox** [reply]
`call(outsideClickHandler)`

**Comment by azinit** [comment]
## Что вам ближе?

> Для статистики
>
> Текущие выводы - см. здесь (https://github.com/feature-sliced/wiki/discussions/12#discussioncomment-527852)

**Reply by azinit** [reply]
> 1. Не стремимся за естественным порядком слов, используем свой формальный
> 2. *namespace? + high context (HC) + low context? (LC) + action (A)*

- `userGet`
- `userMessagesGet`
- `clickOutsideHandle`
- `employeeChangeStatus`

**Reply by azinit** [reply]
> 1. Начинаем с глагола, как в естественном языке
> 2. *event => entities => effect* (условно)

- `getUser`
- `getUserMessages`
- `handleOutsideClick`
- `changeEmployeeStatus`

**Reply by akaia-shadowfox** [reply]
Воздержусь, потому что по моему мнению тут всё неоднозначно. Конкретно для API методов – да, соглашусь что первый метод всё-таки оказался удобнее. А вот на всё остальное у меня несколько альтернативные взгляды

**Reply by azinit** [reply]
> https://github.com/feature-sliced/wiki/discussions/12#discussioncomment-527852
>
> https://github.com/feature-sliced/wiki/discussions/12#discussioncomment-527879

@karina-drummer, про то и писал)
